[MLIR][TOSA] Add --tosa-reduce-transposes pass

arteen1000 opened the pull request:
----------
Motivation:
----------

Some legalization pathways introduce redundant tosa.TRANSPOSE
operations that result in avoidable data movement. For example,
PyTorch -> TOSA contains a lot of unnecessary transposes due
to conversions between NCHW and NHWC.

We wish to remove all the ones that we can, since in general
it is possible to remove the overwhelming majority.

------------
Changes Made:
------------

- Add the --tosa-reduce-transposes pass
- Add TosaElementwiseOperator trait.

-------------------
High-Level Overview:
-------------------

The pass works through the transpose operators in the program. It begins at some
transpose operator with an associated permutations tensor. It traverses upwards
through the dependencies of this transpose and verifies that we encounter only
operators with the TosaElementwiseOperator trait and terminate in either
constants, reshapes, or transposes.

We then evaluate whether there are any additional restrictions (the transposes
it terminates in must invert the one we began at, and the reshapes must be ones
in which we can fold the transpose into), and then we hoist the transpose through
the intervening operators, folding it at the constants, reshapes, and transposes.

Finally, we ensure that we do not need both the transposed form (the form that
had the transpose hoisted through it) and the untransposed form (which it was prior),
by analyzing the usages of those dependent operators of a given transpose we are
attempting to hoist and replace.

If they are such that it would require both forms to be necessary, then we do not
replace the hoisted transpose, causing the new chain to be dead. Otherwise, we do
and the old chain (untransposed form) becomes dead. Only one chain will ever then
be live, resulting in no duplication.

We then perform a simple one-pass DCE, so no canonicalization is necessary.

--------------
Impact of Pass:
--------------

Patching the dense_resource artifacts (from PyTorch) with dense attributes to
permit constant folding, we receive the following results.

Note that data movement represents total transpose data movement, calculated
by noting which dimensions moved during the transpose.

///////////
MobilenetV3:
///////////

BEFORE total data movement: 11798776 B (11.25 MiB)
AFTER total data movement: 2998016 B (2.86 MiB)
74.6% of data movement removed.

BEFORE transposes: 82
AFTER transposes: 20
75.6% of transposes removed.

////////
ResNet18:
////////

BEFORE total data movement: 20596556 B (19.64 MiB)
AFTER total data movement: 1003520 B (0.96 MiB)
95.2% of data movement removed.

BEFORE transposes: 56
AFTER transposes: 5
91.1% of transposes removed.

////////
ResNet50:
////////

BEFORE total data movement: 83236172 B (79.3 MiB)
AFTER total data movement: 3010560 B (2.87 MiB)
96.4% of data movement removed

BEFORE transposes: 120
AFTER transposes: 7
94.2% of transposes removed.

/////////
ResNet101:
/////////

BEFORE total data movement: 124336460 B (118.58 MiB)
AFTER total data movement: 3010560 B (2.87 MiB)
97.6% of data movement removed

BEFORE transposes: 239
AFTER transposes: 7
97.1% of transposes removed.

/////////
ResNet152:
/////////

BEFORE total data movement: 175052108 B (166.94 MiB)
AFTER total data movement: 3010560 B (2.87 MiB)
98.3% of data movement removed

BEFORE transposes: 358
AFTER transposes: 7
98.0% of transposes removed.

////////
Overview:
////////

We see that we remove up to 98% of transposes and eliminate
up to 98.3% of redundant transpose data movement.

In the context of ResNet50, with 120 inferences per second,
we reduce dynamic transpose data bandwidth from 9.29 GiB/s
to 344.4 MiB/s.

-----------
Future Work:
-----------

(1) Evaluate tradeoffs with permitting ConstOp to be duplicated across hoisted
    transposes with different permutation tensors.

(2) Expand the class of foldable upstream ReshapeOp we permit beyond
    N -> 1x1x...x1xNx1x...x1x1.

(3) Enchance the pass to permit folding arbitrary transpose pairs, beyond
    those that form the identity.

(4) Add support for more instructions besides TosaElementwiseOperator as
    the intervening ones (for example, the reduce_* operators).

(5) Support hoisting transposes up to an input parameter.

rsuderman reviewed (approved):
I would recommend merging the test files together. I don't see a lot of value of separating them out from eachother.

joker-eph commented on mlir/include/mlir/Dialect/Tosa/Transforms/Passes.td:
```diff
@@ -126,4 +126,20 @@ def TosaValidation : Pass<"tosa-validate", "mlir::ModuleOp"> {
+    tosa.RESHAPE, a nullifying tosa.TRANSPOSE, or a tosa.CONST. It then propagates the downstream
+    transform upward through the intervening operators if it is able and replaces the downstream tosa.TRANSPOSE.
+    Results generally better when run after Canonicalization and resolution of dynamic shapes.
+    Canonicalization is required for dead code elimination after pass is run.
```
That seems costly, can a cheap DCE be integrated in the pass?

joker-eph commented on mlir/include/mlir/Dialect/Tosa/Transforms/Passes.td:
```diff
@@ -126,4 +126,20 @@ def TosaValidation : Pass<"tosa-validate", "mlir::ModuleOp"> {
+    of data-layout transformations when legalizing to TOSA, a common one being transformations between NHWC and NCHW
+    layouts.
+  }];
+  let constructor = "tosa::createTosaRemoveRedundantTransposes()";
```
Please remove this line, it'll be all auto-generated.

arteen1000 commented:
Modified per request @joker-eph.

arteen1000 committed:
[MLIR][TOSA] Add --tosa-reduce-transposes pass

----------
Motivation:
----------

Some legalization pathways introduce redundant tosa.TRANSPOSE
operations that result in avoidable data movement. For example,
PyTorch -> TOSA contains a lot of unnecessary transposes due
to conversions between NCHW and NHWC.

We wish to remove all the ones that we can, since in general
it is possible to remove the overwhelming majority.

------------
Changes Made:
------------

- Add the --tosa-reduce-transposes pass
- Add TosaElementwiseOperator trait.

-------------------
High-Level Overview:
-------------------

The pass works through the transpose operators in the program. It begins at some
transpose operator with an associated permutations tensor. It traverses upwards
through the dependencies of this transpose and verifies that we encounter only
operators with the TosaElementwiseOperator trait and terminate in either
constants, reshapes, or transposes.

We then evaluate whether there are any additional restrictions (the transposes
it terminates in must invert the one we began at, and the reshapes must be ones
in which we can fold the transpose into), and then we hoist the transpose through
the intervening operators, folding it at the constants, reshapes, and transposes.

Finally, we ensure that we do not need both the transposed form (the form that
had the transpose hoisted through it) and the untransposed form (which it was prior),
by analyzing the usages of those dependent operators of a given transpose we are
attempting to hoist and replace.

If they are such that it would require both forms to be necessary, then we do not
replace the hoisted transpose, causing the new chain to be dead. Otherwise, we do
and the old chain (untransposed form) becomes dead. Only one chain will ever then
be live, resulting in no duplication.

We then perform a simple one-pass DCE, so no canonicalization is necessary.

--------------
Impact of Pass:
--------------

Patching the dense_resource artifacts (from PyTorch) with dense attributes to
permit constant folding, we receive the following results.

Note that data movement represents total transpose data movement, calculated
by noting which dimensions moved during the transpose.

///////////
MobilenetV3:
///////////

BEFORE total data movement: 11798776 B (11.25 MiB)
AFTER total data movement: 2998016 B (2.86 MiB)
74.6% of data movement removed.

BEFORE transposes: 82
AFTER transposes: 20
75.6% of transposes removed.

////////
ResNet18:
////////

BEFORE total data movement: 20596556 B (19.64 MiB)
AFTER total data movement: 1003520 B (0.96 MiB)
95.2% of data movement removed.

BEFORE transposes: 56
AFTER transposes: 5
91.1% of transposes removed.

////////
ResNet50:
////////

BEFORE total data movement: 83236172 B (79.3 MiB)
AFTER total data movement: 3010560 B (2.87 MiB)
96.4% of data movement removed

BEFORE transposes: 120
AFTER transposes: 7
94.2% of transposes removed.

/////////
ResNet101:
/////////

BEFORE total data movement: 124336460 B (118.58 MiB)
AFTER total data movement: 3010560 B (2.87 MiB)
97.6% of data movement removed

BEFORE transposes: 239
AFTER transposes: 7
97.1% of transposes removed.

/////////
ResNet152:
/////////

BEFORE total data movement: 175052108 B (166.94 MiB)
AFTER total data movement: 3010560 B (2.87 MiB)
98.3% of data movement removed

BEFORE transposes: 358
AFTER transposes: 7
98.0% of transposes removed.

////////
Overview:
////////

We see that we remove up to 98% of transposes and eliminate
up to 98.3% of redundant transpose data movement.

In the context of ResNet50, with 120 inferences per second,
we reduce dynamic transpose data bandwidth from 9.29 GiB/s
to 344.4 MiB/s.

-----------
Future Work:
-----------

(1) Evaluate tradeoffs with permitting ConstOp to be duplicated across hoisted
    transposes with different permutation tensors.

(2) Expand the class of foldable upstream ReshapeOp we permit beyond
    N -> 1x1x...x1xNx1x...x1x1.

(3) Enchance the pass to permit folding arbitrary transpose pairs, beyond
    those that form the identity.

(4) Add support for more instructions besides TosaElementwiseOperator as
    the intervening ones (for example, the reduce_* operators).

(5) Support hoisting transposes up to an input parameter.

Signed-off-by: Arteen Abrishami <arteen.abrishami@arm.com>

sjarus commented:
@arteen1000 is away from desk and had collected this data. 'Data movement' below refers to transposed data, counting only axes that are transposed, of all the tosa.transpose() instances in the IR. These tests were done as part of functional and performance experiments on Torch-MLIR to validate that the final results are functionally unaffected. 

///////////
MobilenetV3:
///////////

BEFORE total data movement: 11798776 B (11.25 MiB)
AFTER total data movement: 2998016 B (2.86 MiB)
74.6% of data movement removed.

BEFORE transposes: 82
AFTER transposes: 20
75.6% of transposes removed.

////////
ResNet18:
////////

BEFORE total data movement: 20596556 B (19.64 MiB)
AFTER total data movement: 1003520 B (0.96 MiB)
95.2% of data movement removed.

BEFORE transposes: 56
AFTER transposes: 5
91.1% of transposes removed.

////////
ResNet50:
////////

BEFORE total data movement: 83236172 B (79.3 MiB)
AFTER total data movement: 3010560 B (2.87 MiB)
96.4% of data movement removed

BEFORE transposes: 120
AFTER transposes: 7
94.2% of transposes removed.

/////////
ResNet101:
/////////

BEFORE total data movement: 124336460 B (118.58 MiB)
AFTER total data movement: 3010560 B (2.87 MiB)
97.6% of data movement removed

BEFORE transposes: 239
AFTER transposes: 7
97.1% of transposes removed.

/////////
ResNet152:
/////////

BEFORE total data movement: 175052108 B (166.94 MiB)
AFTER total data movement: 3010560 B (2.87 MiB)
98.3% of data movement removed

BEFORE transposes: 358
AFTER transposes: 7
98.0% of transposes removed.

jpienaar commented:
Funnily I've seen two other teams do almost the same thing recently. @cathyzhyi you have some context of this pass in case you could take a look too.

joker-eph reviewed (commented):
Just some drive-by comments, I only skimmed the code: kudos for the great comments all along!

sjarus commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,733 @@
+
+void TosaRemoveRedundantTransposes::runOnOperation() {
+  // We want to operate only within a single block.
+  // Call --inline before to run the pass.
```
This is probably a stale comment - it's leftover from early conversation on TBD work on control flow ops that have nested region bodies and other such constructs. Would removing this comment resolve this, @joker-eph ?

eric-k256 reviewed (approved):
Looks good to me. Thanks!

jpienaar commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+    ArrayRef<int32_t> perms1, ArrayRef<int32_t> perms2) {
+  if (perms1.size() != perms2.size())
+    return false;
+  for (int32_t i = 0; i < static_cast<int32_t>(perms1.size()); i++)
```
https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

jpienaar replied:
(in this case it would seem a range based loop can be used)

arteen1000 replied:
Will evaluate prior. Need the `i` in the for loop, unless I am missing something.

jpienaar commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+  // these interveaning operations. For example,
+  // %0 = tosa.clamp %input : (tensor<2x3xi32>) -> tensor<2x3xi32>
+  // %1 = tosa.transpose %0 {perms = [1, 0]} : (tensor<2x3xi32>) ->
+  // tensor<3x2xi32> becomes: %0 = tosa.transpose %input {perms = [1, 0]} :
```
Nit: make the IR on its own lines with become in between

jpienaar commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+  // from this new chain.
+  return rewriter
+      .create(op->getLoc(),
+              rewriter.getStringAttr(op->getName().getStringRef()), operands,
```
It shouldn't be needed to create a new string attribute here.

jpienaar commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+  auto reshapeOutputType =
+      llvm::cast<RankedTensorType>(reshapeOutput.getType());
+
+  // Instead of inserting a TransposeOp here, we
```
Could you reflow this?

jpienaar commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+    rewriter.setInsertionPointAfter(op);
+
+    std::optional<Value> maybeValue =
+        llvm::TypeSwitch<Operation *, std::optional<Value>>(op)
```
I think you can just do

.Case<ConstOp, ReshapeOp, TransposeOp>([&](auto op) { return ... }

jpienaar commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+        if (failed(otherTranspose.getConstantPerms(otherPerms)) ||
+            !llvm::equal(perms, otherPerms))
+          return false;
+
```
Nit: remove empty line here.

jpienaar commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+            !llvm::equal(perms, otherPerms))
+          return false;
+
+      } else if (llvm::none_of(
```
This uses a lot of white space. Maybe hoist out into helper.

jpienaar commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+    collectedPerms.emplace_back();
+    SmallVector<int32_t> &perms = collectedPerms.back();
+
+    // Dynamic shapes are OK,
```
Reflow

jpienaar commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+  // TransposeOp and their perms, a necessity for the final replacements.
+  std::stack<std::pair<TransposeOp, ArrayRef<int32_t>>> totalTransposeOrder;
+
+  // We want to reserve the space up front,
```
Reflow comment

arteen1000 commented on mlir/include/mlir/Dialect/Tosa/IR/TosaOpBase.td:
```diff
@@ -206,6 +206,15 @@ def Tosa_ExplicitValuePadOpQuantInfoBuilder : OpBuilder<
+// TOSA Operator Trait.
+//===----------------------------------------------------------------------===//
+
+// Permits broadcasting. Elementwise trait is too strict.
```
The element-wise operators in TOSA all have broadcasting semantics, so we wouldn't ever use `Elementwise`.
As an example, we want to allow something like `tosa.add %0, %1 : (tensor<3xi32>, tensor<1xi32>)` which the `Elementwise` trait disallows. For unary, this makes no difference.

arteen1000 commented on mlir/include/mlir/Dialect/Tosa/Transforms/Passes.td:
```diff
@@ -126,4 +126,18 @@ def TosaValidation : Pass<"tosa-validate", "mlir::ModuleOp"> {
+def TosaRemoveRedundantTransposes : Pass<"tosa-remove-redundant-transposes", "func::FuncOp"> {
+  let summary = "Remove redundant transposes";
+  let description = [{
+    Pass that identifies and removes redundant tosa.TRANSPOSE operations.
```
I use the term redundant to mean 'not necessary.'
As an example, the first test `@test_transpose_tracks_to_nullifying_single_step`, you can see how those transposes might be unnecessary.

arteen1000 commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+// this chain we bubbled-up into its transposed form. We do this by analyzing
+// the dependency fan-ins across all transposes with the same perms tensor
+// in order to ensure that they do not have uses outside this group, which
+// would cause the old code section to remain "live", and not removed by
```
We would increase the number of operators significantly, since we would then have potentially two chains, one in transposed form and one in untransposed form. In ResNet and MobileNet this was noted to be a problem.

If you have
```
transpose
... long chain ...
transpose
```
and some operator in that chain has a usage outside of that chain, then we would need both the untransposed and transposed forms of the operators in that chain.

arteen1000 commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+// High-Level Overview:
+// -------------------
+
+// The pass begins at a downstream transpose with some perms tensor.
```
I used `perms` because that is what the `tosa.TRANSPOSE` operator argument calls it, so I was aligning with that.

cathyzhyi commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+
+// (3)
+
+// Make the pass more general, beyond just allowing upstream transposes
```
nit: I feel this paragraph could be more concise to be something like: Enhance the pass's generality by enabling it to bubble up and merge a downstream transpose with upstream transposes, even when the resulting transpose is not an identity transformation.

arteen1000 commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+//===- TosaRemoveRedundantTransposes.cpp
```
This is a problem with `clang-format`. I have `clang-format` on-save. I just attempted to fix it without formatting, ran `git clang-format HEAD~1` and it failed the check.

cathyzhyi commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+  // since generally we can't propagate a TransposeOp through it.
+  // TransposeOp, ReshapeOp, ConstOp will have no in-edges in the data
+  // dependency graph we construct for the downstream TransposeOp.
+  if (!llvm::isa<TransposeOp>(op) && !llvm::isa<ReshapeOp>(op) &&
```
nit: easier to read to check for `llvm::isa<TransposeOp>(op) || llvm::isa<ReshapeOp>(op) ||  llvm::isa<ConstOp>(op)` and return

arteen1000 replied:
We do need to insert them, which occurs where it says `collected.insert(op)`, which is why it does it this way.

arteen1000 commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+
+// Support bubbling transposes up to the input parameter. May not
+// need extensive fan-in analysis as no operation cost associated
+// if used elsewhere.
```
For example, let's say I have a construct:
```
%0 = transpose %arg1
%1 = tosa.add %arg0, %0
%2 = transpose %1
```
then I could transform this to
```
%0 = transpose %arg0
%1 = tosa.add %0, %arg1
```
Thereby permitting us other "upstream" transposes.

cathyzhyi replied:
I see but this looks the same as the current one where a down stream transpose bubbles up and cancel out with the upstream transposes.

arteen1000 replied:
Like mentioned, the PR currently requires that the upstream dependencies of the transpose be either constants, transposes, or reshapes -- with the ones that come between those and our downstream transpose being elementwise. There's some good examples in the test file. In this case, when we go up the transpose and go up the add, we hit a block argument -- which doesn't fit that criteria. If %arg0 was the result of some operator that eventually terminated in constants, inverse transposes, or the class of reshapes we cover, then it would be handled. The reason it's a future work is because the networks we encountered don't have this direct argument use like the above.

cathyzhyi replied:
I see. Thanks for the detailed explanation.

arteen1000 commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+  // Asserted by TransposeOp verifier and TOSA disallowing tensor with dimension
+  // 0.
+  // If not in place, something is very wrong.
+  if (rank <= 0 || oldType.getNumElements() <= 0 || perms.size() != rank) {
```
Thanks. I'd placed it there because those assumptions are actually important for what comes after.

jpienaar reviewed (approved):
This looks good as start. Remember to update the PR description before landing (and just regular run through clang-tidy and clang-format to catch small changes)

sjarus commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+// as the sources of the data dependencies, and TosaElementWiseOperator
+// after that, if the function returns true.
+bool TosaRemoveRedundantTransposes::collectFanIn(
+    Operation *op, SetVector<Operation *> &collected) {
```
Yes, due to the restricted supported op set this fundamentally a filtered getBackwardSlice() . Noted as follow on work.

sjarus commented on mlir/lib/Dialect/Tosa/Transforms/TosaRemoveRedundantTransposes.cpp:
```diff
@@ -0,0 +1,731 @@
+
+    for (Value operand : op->getOperands()) {
+
+      if (!collectFanIn(operand.getDefiningOp(), collected))
```
Noted as follow-on performance work as agreed.

sjarus commented:
With the concurrence of Rob, Mehdi and Jacques I'm going to proceed and land this. Thanks for this work, @arteen1000 !
